src/Storages/Cache/RemoteCacheController.cpp

```diff
@@ -0,0 +1,249 @@
+{
+namespace fs = std::filesystem;
+namespace ErrorCodes
+{
+    extern const int OK;
+    extern const int BAD_ARGUMENTS;
+    extern const int LOGICAL_ERROR;
+    extern const int END_OF_FILE;
+}
+
+bool RemoteCacheController::loadInnerInformation(const fs::path & file_path)
+{
+    if (!fs::exists(file_path))
+        return false;
+    std::ifstream info_file(file_path);
+    Poco::JSON::Parser info_parser;
+    auto info_json = info_parser.parse(info_file).extract<Poco::JSON::Object::Ptr>();
+    file_status = static_cast<LocalFileStatus>(info_json->get("file_status").convert<Int32>());
+    metadata_class = info_json->get("metadata_class").convert<String>();
+    info_file.close();
+    return true;
+}
+
+std::shared_ptr<RemoteCacheController> RemoteCacheController::recover(const std::filesystem::path & local_path_)
+{
+    auto * log = &Poco::Logger::get("RemoteCacheController");
+
+    if (!std::filesystem::exists(local_path_ / "data.bin"))
+    {
+        LOG_TRACE(log, "Invalid cached directory:{}", local_path_.string());
+        return nullptr;
+    }
+
+    auto cache_controller = std::make_shared<RemoteCacheController>(nullptr, local_path_, 0);
+    if (!cache_controller->loadInnerInformation(local_path_ / "info.txt")
+            || cache_controller->file_status != DOWNLOADED)
+    {
+        LOG_INFO(log, "Recover cached file failed. local path:{}", local_path_.string());
+        return nullptr;
+    }
+    try
+    {
+        cache_controller->file_metadata_ptr = RemoteFileMetadataFactory::instance().get(cache_controller->metadata_class);
+    }
+    catch(const Exception & e)
+    {
+        LOG_ERROR(log, "Get metadata class failed for {}. {}", cache_controller->metadata_class, e.message());
+        cache_controller->file_metadata_ptr = nullptr;
+    }
+    if (!cache_controller->file_metadata_ptr)
+    {
+        // do not load this invalid cached file and clear it. the clear action is in
+        // ExternalDataSourceCache::recoverCachedFilesMetadata(), because deleting directories during iteration will
+        // cause unexpected behaviors
+        LOG_ERROR(log, "Cannot create the metadata class : {}. The cached file is invalid and will be remove. path:{}",
+                cache_controller->metadata_class,
+                local_path_.string());
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Invalid metadata class:{}", cache_controller->metadata_class);
+    }
+    std::ifstream metadata_file(local_path_ / "metadata.txt");
+    if (!cache_controller->file_metadata_ptr->fromString(std::string((std::istreambuf_iterator<char>(metadata_file)),
+                    std::istreambuf_iterator<char>())))
+    {
+        LOG_ERROR(log, "Cannot load the metadata. The cached file is invalid and will be remove. path:{}",
+                local_path_.string());
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Invalid metadata file({}) for meta class {}", local_path_.string(), cache_controller->metadata_class);
```

Comment:
I think better to remove error logging if you anyway throw exception.

Reply:
done